In [52]:
import numpy as np
import pandas as pd 
import seaborn as sns 
import plotly.graph_objects as go

In [5]:
app_open = pd.read_csv(r'C:\Users\Home\Desktop\DS_Projects\task_1_subscription_funnel\app_open_data.csv')

In [6]:
app_open = app_open.drop('mycol', axis = 1)

In [7]:
app_open.head(10)

,event_date,timestamp,device_skey,session_skey,user_skey,is_first_app_open,platform,country_code
0,2023-11-13,2023-11-13T13:15:04.155Z,-1160721689826509359,-6069694996621106477,-4198274634774068263,False,android,vn
1,2023-11-13,2023-11-13T11:32:23.553Z,-4659317238696778303,2034623284200851906,-8645779688131645005,False,android,vn
2,2023-11-13,2023-11-13T02:00:57.992Z,8574209580357038155,7700760069378643919,3858142552250413010,False,android,br
3,2023-11-13,2023-11-13T09:39:29.349Z,-4107400260788282015,-3321024403238710334,3858142552250413010,False,android,br
4,2023-11-13,2023-11-13T07:45:28.949Z,-6656369459599123810,-4504039664568204955,8414965012823618631,False,apple,vn
5,2023-11-13,2023-11-13T05:13:23.507Z,7512342648113163912,7470111580885362241,-8945100976950520118,False,android,vn
6,2023-11-13,2023-11-13T21:18:15.461Z,-3472619401995420955,5574459885572744489,1680094967225068296,False,android,br
7,2023-11-13,2023-11-13T22:04:13.426Z,-4548147938051377057,915445541882277202,6598616922021353082,False,android,eg
8,2023-11-13,2023-11-13T17:37:20.487Z,7557963256979044323,-5454928798009981395,7159509098064362553,False,android,eg
9,2023-11-13,2023-11-13T08:30:12.736Z,2071370451767828968,-4582424473293925217,6133534407697871217,False,apple,vn


In [8]:
app_open["timestamp"] = pd.to_datetime(app_open["timestamp"], utc = True)

In [9]:
app_open

,event_date,timestamp,device_skey,session_skey,user_skey,is_first_app_open,platform,country_code
0,2023-11-13,2023-11-13 13:15:04.155000+00:00,-1160721689826509359,-6069694996621106477,-4198274634774068263,False,android,vn
1,2023-11-13,2023-11-13 11:32:23.553000+00:00,-4659317238696778303,2034623284200851906,-8645779688131645005,False,android,vn
2,2023-11-13,2023-11-13 02:00:57.992000+00:00,8574209580357038155,7700760069378643919,3858142552250413010,False,android,br
3,2023-11-13,2023-11-13 09:39:29.349000+00:00,-4107400260788282015,-3321024403238710334,3858142552250413010,False,android,br
4,2023-11-13,2023-11-13 07:45:28.949000+00:00,-6656369459599123810,-4504039664568204955,8414965012823618631,False,apple,vn
...,...,...,...,...,...,...,...,...
39208,2023-11-07,2023-11-07 17:45:28.534000+00:00,-1621212041562007236,-22542611557668742,6419637895506032089,False,apple,de
39209,2023-11-07,2023-11-07 18:22:23.558000+00:00,-152653709468642127,1515180927234277919,9082017717387193756,False,android,br
39210,2023-11-07,2023-11-07 09:50:22.298000+00:00,-8441555556049786810,-3353746405010624321,2041568181486525324,False,apple,vn
39211,2023-11-07,2023-11-07 14:19:43.979000+00:00,3559314631456246419,7529333791297578561,-6293457713645842564,False,android,vn


In [10]:
app_open['event_name'] = 'app_open'

In [11]:
app_open

,event_date,timestamp,device_skey,session_skey,user_skey,is_first_app_open,platform,country_code,event_name
0,2023-11-13,2023-11-13 13:15:04.155000+00:00,-1160721689826509359,-6069694996621106477,-4198274634774068263,False,android,vn,app_open
1,2023-11-13,2023-11-13 11:32:23.553000+00:00,-4659317238696778303,2034623284200851906,-8645779688131645005,False,android,vn,app_open
2,2023-11-13,2023-11-13 02:00:57.992000+00:00,8574209580357038155,7700760069378643919,3858142552250413010,False,android,br,app_open
3,2023-11-13,2023-11-13 09:39:29.349000+00:00,-4107400260788282015,-3321024403238710334,3858142552250413010,False,android,br,app_open
4,2023-11-13,2023-11-13 07:45:28.949000+00:00,-6656369459599123810,-4504039664568204955,8414965012823618631,False,apple,vn,app_open
...,...,...,...,...,...,...,...,...,...
39208,2023-11-07,2023-11-07 17:45:28.534000+00:00,-1621212041562007236,-22542611557668742,6419637895506032089,False,apple,de,app_open
39209,2023-11-07,2023-11-07 18:22:23.558000+00:00,-152653709468642127,1515180927234277919,9082017717387193756,False,android,br,app_open
39210,2023-11-07,2023-11-07 09:50:22.298000+00:00,-8441555556049786810,-3353746405010624321,2041568181486525324,False,apple,vn,app_open
39211,2023-11-07,2023-11-07 14:19:43.979000+00:00,3559314631456246419,7529333791297578561,-6293457713645842564,False,android,vn,app_open


In [12]:
app_open.info()

<class 'pandas.DataFrame'>
RangeIndex: 39213 entries, 0 to 39212
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype              
---  ------             --------------  -----              
 0   event_date         39213 non-null  str                
 1   timestamp          39213 non-null  datetime64[us, UTC]
 2   device_skey        39213 non-null  int64              
 3   session_skey       39213 non-null  int64              
 4   user_skey          39213 non-null  int64              
 5   is_first_app_open  39213 non-null  bool               
 6   platform           39213 non-null  str                
 7   country_code       39213 non-null  str                
 8   event_name         39213 non-null  str                
dtypes: bool(1), datetime64[us, UTC](1), int64(3), str(4)
memory usage: 2.4 MB


In [13]:
event_data = pd.read_csv(r'C:\Users\Home\Desktop\DS_Projects\task_1_subscription_funnel\event_data.csv')

In [14]:
event_data.head(10)

,event_name,timestamp,event_skey,device_skey,user_skey,session_skey,platform,source,country_code
0,registration_open,1699782597828,2040623284347280262,-4243964682902975820,NaN,2.275135e+18,android,user_profile,vn
1,registration_open,1699630836825,-2687516005135594704,4342432628889352902,NaN,-4.527749e+18,apple,app_start,de
2,registration_open,1699708549187,-4765262403246217267,6708321608618278911,NaN,-2.576991e+18,apple,app_start,eg
3,registration_open,1699710702104,-2105604357975282211,-3303323474515376120,NaN,8.616913e+18,android,user_profile,br
4,registration_open,1700064536040,4905416007576559228,1875496956373071881,NaN,6.956769e+18,android,app_start,eg
5,registration_open,1700066972678,6030010734792124838,1983898266439418900,NaN,-6.472438e+17,android,app_start,vn
6,registration_open,1700068588750,-211389582663558542,7619586686974168999,NaN,1.983699e+18,android,app_start,eg
7,registration_open,1700068596488,1161442037272662347,7619586686974168999,NaN,1.983699e+18,android,registration,eg
8,registration_open,1700068600764,4756048876717438951,7619586686974168999,NaN,1.983699e+18,android,registration,eg
9,registration_open,1699712848047,-2079815722091521874,-3827571418738237435,NaN,1.209165e+18,android,user_profile,br


In [15]:
event_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 151947 entries, 0 to 151946
Data columns (total 9 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   event_name    151947 non-null  str    
 1   timestamp     151947 non-null  int64  
 2   event_skey    151947 non-null  int64  
 3   device_skey   151947 non-null  int64  
 4   user_skey     113649 non-null  float64
 5   session_skey  150752 non-null  float64
 6   platform      151947 non-null  str    
 7   source        146479 non-null  str    
 8   country_code  151947 non-null  str    
dtypes: float64(2), int64(3), str(4)
memory usage: 10.4 MB


In [16]:
event_data.isnull().sum()

event_name          0
timestamp           0
event_skey          0
device_skey         0
user_skey       38298
session_skey     1195
platform            0
source           5468
country_code        0
dtype: int64

In [17]:
event_data['timestamp'] = pd.to_datetime(event_data['timestamp'], unit = 'ms', utc = True)

In [18]:
event_data.head()

,event_name,timestamp,event_skey,device_skey,user_skey,session_skey,platform,source,country_code
0,registration_open,2023-11-12 09:49:57.828000+00:00,2040623284347280262,-4243964682902975820,NaN,2.275135e+18,android,user_profile,vn
1,registration_open,2023-11-10 15:40:36.825000+00:00,-2687516005135594704,4342432628889352902,NaN,-4.527749e+18,apple,app_start,de
2,registration_open,2023-11-11 13:15:49.187000+00:00,-4765262403246217267,6708321608618278911,NaN,-2.576991e+18,apple,app_start,eg
3,registration_open,2023-11-11 13:51:42.104000+00:00,-2105604357975282211,-3303323474515376120,NaN,8.616913e+18,android,user_profile,br
4,registration_open,2023-11-15 16:08:56.040000+00:00,4905416007576559228,1875496956373071881,NaN,6.956769e+18,android,app_start,eg


In [19]:
app_open.head()

,event_date,timestamp,device_skey,session_skey,user_skey,is_first_app_open,platform,country_code,event_name
0,2023-11-13,2023-11-13 13:15:04.155000+00:00,-1160721689826509359,-6069694996621106477,-4198274634774068263,False,android,vn,app_open
1,2023-11-13,2023-11-13 11:32:23.553000+00:00,-4659317238696778303,2034623284200851906,-8645779688131645005,False,android,vn,app_open
2,2023-11-13,2023-11-13 02:00:57.992000+00:00,8574209580357038155,7700760069378643919,3858142552250413010,False,android,br,app_open
3,2023-11-13,2023-11-13 09:39:29.349000+00:00,-4107400260788282015,-3321024403238710334,3858142552250413010,False,android,br,app_open
4,2023-11-13,2023-11-13 07:45:28.949000+00:00,-6656369459599123810,-4504039664568204955,8414965012823618631,False,apple,vn,app_open


In [20]:
unified_df = pd.concat([app_open, event_data], ignore_index=True)

In [21]:
unified_df.head(10)

,event_date,timestamp,device_skey,session_skey,user_skey,is_first_app_open,platform,country_code,event_name,event_skey,source
0,2023-11-13,2023-11-13 13:15:04.155000+00:00,-1160721689826509359,-6.069695e+18,-4.198275e+18,False,android,vn,app_open,NaN,NaN
1,2023-11-13,2023-11-13 11:32:23.553000+00:00,-4659317238696778303,2.034623e+18,-8.645780e+18,False,android,vn,app_open,NaN,NaN
2,2023-11-13,2023-11-13 02:00:57.992000+00:00,8574209580357038155,7.700760e+18,3.858143e+18,False,android,br,app_open,NaN,NaN
3,2023-11-13,2023-11-13 09:39:29.349000+00:00,-4107400260788282015,-3.321024e+18,3.858143e+18,False,android,br,app_open,NaN,NaN
4,2023-11-13,2023-11-13 07:45:28.949000+00:00,-6656369459599123810,-4.504040e+18,8.414965e+18,False,apple,vn,app_open,NaN,NaN
5,2023-11-13,2023-11-13 05:13:23.507000+00:00,7512342648113163912,7.470112e+18,-8.945101e+18,False,android,vn,app_open,NaN,NaN
6,2023-11-13,2023-11-13 21:18:15.461000+00:00,-3472619401995420955,5.574460e+18,1.680095e+18,False,android,br,app_open,NaN,NaN
7,2023-11-13,2023-11-13 22:04:13.426000+00:00,-4548147938051377057,9.154455e+17,6.598617e+18,False,android,eg,app_open,NaN,NaN
8,2023-11-13,2023-11-13 17:37:20.487000+00:00,7557963256979044323,-5.454929e+18,7.159509e+18,False,android,eg,app_open,NaN,NaN
9,2023-11-13,2023-11-13 08:30:12.736000+00:00,2071370451767828968,-4.582424e+18,6.133534e+18,False,apple,vn,app_open,NaN,NaN


In [22]:
unified_df = unified_df.sort_values(by=['device_skey', 'timestamp']).reset_index(drop=True)  
unified_df.head(10)

,event_date,timestamp,device_skey,session_skey,user_skey,is_first_app_open,platform,country_code,event_name,event_skey,source
0,NaN,2023-11-16 21:35:44.221000+00:00,-9223169578178423456,4.840357e+18,NaN,NaN,android,de,editor_open,-4.166152e+18,editor
1,NaN,2023-11-13 23:53:00.865000+00:00,-9223109751360149320,-2.786915e+18,-5.232895e+18,NaN,android,br,editor_open,-1.785653e+18,history_player
2,NaN,2023-11-13 23:57:01.420000+00:00,-9223109751360149320,-2.786915e+18,-5.232895e+18,NaN,android,br,subscription_offer_open,2.751535e+18,history_player_add_text
3,NaN,2023-11-13 23:57:08.014000+00:00,-9223109751360149320,-2.786915e+18,-5.232895e+18,NaN,android,br,subscription_offer_open,5.418239e+18,history_player_add_text
4,NaN,2023-11-13 23:57:34.911000+00:00,-9223109751360149320,-2.786915e+18,-5.232895e+18,NaN,android,br,object_export,-7.474779e+18,editor_screen
5,NaN,2023-11-13 23:57:42.834000+00:00,-9223109751360149320,-2.786915e+18,-5.232895e+18,NaN,android,br,object_export,-5.156694e+17,export_screen
6,NaN,2023-11-13 23:58:38.566000+00:00,-9223109751360149320,-2.786915e+18,-5.232895e+18,NaN,android,br,editor_open,-1.574190e+18,my_network_inspiring
7,NaN,2023-11-09 22:23:16.567000+00:00,-9223104094521336251,8.656567e+18,-2.222050e+18,NaN,android,eg,create_flow_open,-3.506750e+18,my_network
8,NaN,2023-11-09 22:23:23.552000+00:00,-9223104094521336251,8.656567e+18,-2.222050e+18,NaN,android,eg,editor_open,-2.531039e+18,create_flow
9,NaN,2023-11-09 22:23:33.687000+00:00,-9223104094521336251,8.656567e+18,-2.222050e+18,NaN,android,eg,object_export,-3.243063e+18,editor_square_fit


FUNNEL EXTRACTION

In [51]:
# 1. EXPLICITLY enforce the timeline order in memory
funnel_stages = [
    'app_open', 
    'registration_open', 
    'registration_done', 
    'create_flow_open', 
    'editor_open', 
    'subscription_offer_open', 
    'object_export', 
    'subscription_done'
]

# 2. Extract counts using this strict order
funnel_counts = []
for stage in funnel_stages:
    stage_data = unified_df[unified_df['event_name'] == stage]
    unique_devices = stage_data['device_skey'].nunique()
    funnel_counts.append({'stage': stage, 'unique_devices': unique_devices})

overall_funnel = pd.DataFrame(funnel_counts)

# 3. Calculate conversions safely
overall_funnel['stage_to_stage_conv_%'] = (overall_funnel['unique_devices'] / overall_funnel['unique_devices'].shift(1)) * 100
overall_funnel['stage_to_stage_conv_%'] = overall_funnel['stage_to_stage_conv_%'].fillna(100.0).round(2)

starting_volume = overall_funnel.loc[0, 'unique_devices'] # Now guaranteed to be app_open
overall_funnel['total_conv_from_start_%'] = (overall_funnel['unique_devices'] / starting_volume) * 100
overall_funnel['total_conv_from_start_%'] = overall_funnel['total_conv_from_start_%'].round(2)

overall_funnel

,stage,unique_devices,stage_to_stage_conv_%,total_conv_from_start_%
0,app_open,38730,100.00,100.00
1,registration_open,4099,10.58,10.58
2,registration_done,2300,56.11,5.94
3,create_flow_open,24681,1073.09,63.73
4,editor_open,25914,105.00,66.91
5,subscription_offer_open,11451,44.19,29.57
6,object_export,20115,175.66,51.94
7,subscription_done,92,0.46,0.24


Architectural Decision: Unit of Analysis
For the funnel construction, device_skey was strictly utilized as the primary unit of analysis rather than user_skey. Exploration of the data reveals that user_skey is inherently NaN for approximately 99% of top-of-funnel events (such as registration_open). Because users exist in an anonymous state prior to completing registration, attempting to track the funnel via user_skey would artificially truncate the data and exclude the vast majority of app opens. By using device_skey, we maintain a continuous, unbreakable entity thread from the initial app open through to the final subscription purchase, preserving the mathematical integrity of the funnel.

In [55]:
unified_df['platform'].value_counts()

platform
android    116777
apple       74383
Name: count, dtype: int64

In [56]:
unified_df['country_code'].value_counts()

country_code
br    62537
vn    60438
eg    47572
de    20613
Name: count, dtype: int64

In [53]:
# --- 1. THE SEGMENTATION ENGINE ---
def build_segment_funnel(df, segment_col):
    """
    Dynamically groups the unified event stream by a segment (e.g., platform) 
    and extracts the strict chronological funnel volumes.
    """
    segment_data = []
    # Drop unknown segments so our charts stay clean
    segments = [s for s in df[segment_col].unique() if pd.notna(s) and s != 'unknown']
    
    for seg in segments:
        seg_df = df[df[segment_col] == seg]
        counts = []
        for stage in funnel_stages:
            # Count unique devices per stage for this specific segment
            vol = seg_df[seg_df['event_name'] == stage]['device_skey'].nunique()
            counts.append(vol)
            
        temp_df = pd.DataFrame({'stage': funnel_stages, 'unique_devices': counts, 'segment': seg})
        segment_data.append(temp_df)
        
    return pd.concat(segment_data, ignore_index=True)

# Generate the raw mathematical tables
platform_funnel = build_segment_funnel(unified_df, 'platform')
country_funnel = build_segment_funnel(unified_df, 'country_code')



# VISUALIZATION A: Platform Comparison
fig_platform = go.Figure()

for platform in platform_funnel['segment'].unique():
    df_plat = platform_funnel[platform_funnel['segment'] == platform]
    fig_platform.add_trace(go.Funnel(
        name=platform.capitalize(),
        y=df_plat['stage'],
        x=df_plat['unique_devices'],
        textinfo="value+percent previous", # Automatically calculates the drop-off!
        opacity=0.85
    ))

fig_platform.update_layout(
    title_text="<b>Funnel Conversion by Platform: Android vs. Apple</b><br><sup>Apple users typically display higher end-of-funnel purchasing power. Does the data prove it?</sup>",
    title_x=0.5,
    funnelmode="group"
)
fig_platform.show()

# VISUALIZATION B: Country Comparison
fig_country = go.Figure()

for country in country_funnel['segment'].unique():
    df_count = country_funnel[country_funnel['segment'] == country]
    fig_country.add_trace(go.Funnel(
        name=country.upper(),
        y=df_count['stage'],
        x=df_count['unique_devices'],
        textinfo="value+percent previous",
        opacity=0.85
    ))

fig_country.update_layout(
    title_text="<b>Funnel Conversion by Country (BR, VN, EG, DE)</b><br><sup>Observing localized friction across the user journey.</sup>",
    title_x=0.5,
    funnelmode="group"
)
fig_country.show()

📊 Deliverable 1 Insights: Funnel Drop-Off & Product Hypotheses
Insight 1: The "Guest Mode" Dominance (Registration Drop-Off)
The Data: Only ~10.5% of users (4,099 / 38,730) open the registration screen, and roughly half of those abandon it before completing. Meanwhile, over 24,000 users successfully open the creator flow.
The Hypothesis: The app’s architecture heavily favors a "Guest Mode" experience. Because users can immediately access core editing tools without an account, there is zero immediate incentive to register. For the 10.5% who do attempt it, the ~44% abandonment rate suggests high UX friction (e.g., too many form fields, lack of quick Social Logins like Google/Apple, or an unoptimized UI).

Insight 2: The Premium Value Proposition Gap (The Primary Drop-Off)
The Data: The most severe drop-off in the entire funnel occurs at the paywall. While 11,451 users trigger the subscription offer, only 92 successfully complete the purchase—a catastrophic 0.8% stage-to-stage conversion rate. Concurrently, over 20,000 users successfully export photos.
The Hypothesis: The free tier is actively cannibalizing the premium tier. Because the baseline free tools are robust enough to satisfy the average user's export needs, the perceived value of the premium features (HD portrait, object removal) does not justify the subscription price. Alternatively, the payment gateway itself may have severe localization friction in our target markets (BR, VN, EG, DE), such as lacking preferred local payment methods, causing high-intent users to bounce at checkout. 

Insight 3: Platform Conversion Efficiency (Apple vs. Android)
The Data: While Android dominates top-of-funnel volume (~116K total events), Apple demonstrates superior conversion efficiency at the bottom of the funnel. Despite having significantly fewer users trigger the subscription_offer_open event, Apple users represent a larger absolute volume of subscription_done events (59 Apple vs. 33 Android).
The Hypothesis: Apple (iOS) users historically exhibit a higher willingness to pay for premium creative tools. Additionally, Apple's native in-app purchasing ecosystem (FaceID/Double Click to Pay) often presents less friction than Android's varied payment gateways, leading to a higher completion rate once the paywall is hit.

Insight 4: Geographic Purchasing Power (The Germany Anomaly)
The Data: Germany (DE) has the smallest overall user base (~20K events) compared to the emerging markets (BR, VN, EG). However, looking at the localized funnel, Germany outperforms these larger markets in late-stage conversion relative to its size.
The Hypothesis: This reflects macroeconomic purchasing power. The subscription price point is likely highly affordable for the German market, whereas it represents a significant financial barrier in Brazil, Vietnam, or Egypt. This indicates an immediate need for localized pricing strategies (Purchasing Power Parity) to capture revenue in those high-volume emerging markets.